components/datetime/src/scaffold/calendar.rs

```diff
@@ -303,29 +303,41 @@ impl FormattableAnyCalendarKind {
     }
 
     pub(crate) fn from_preferences(mut prefs: DateTimeFormatterPreferences) -> Self {
-        if let Some(algo) = prefs.calendar_algorithm {
-            if let Ok(kind) = AnyCalendarKind::try_from(algo) {
-                if let Some(res) = Self::try_from_any_calendar_kind(kind) {
-                    return res;
-                }
-            }
+        if let Some(res) = Self::try_from_any_calendar_kind(AnyCalendarKind::new((&prefs).into())) {
+            return res;
         }
-        // Calendar not specified or not supported by DateTimeFormatter
+
+        // Calendar not supported by DateTimeFormatter
         // Currently this is CalendarAlgorithm::Iso8601, CalendarAlgorithm::Hijri(Rgsa)
         // Let AnyCalendarKind constructor select an appropriate fallback
         prefs.calendar_algorithm = None;
-        let kind = AnyCalendarKind::new((&prefs).into());
-        match Self::try_from_any_calendar_kind(kind) {
-            Some(res) => res,
-            None => {
-                debug_assert!(false, "all locale-default calendars are supported");
-                // fall back to something non-Gregorian to make errors more obvious
-                FormattableAnyCalendarKind::Coptic
-            }
+        if let Some(res) = Self::try_from_any_calendar_kind(AnyCalendarKind::new((&prefs).into())) {
+            return res;
         }
+
+        debug_assert!(false, "all locale-default calendars are supported");
+        // fall back to something non-Gregorian to make errors more obvious
+        FormattableAnyCalendarKind::Coptic
     }
 }
 
+#[test]
+fn asfd() {
+    use icu_locale_core::locale;
+    assert_eq!(
+        FormattableAnyCalendarKind::from_preferences(locale!("en-TH-u-ca-iso8601").into()),
+        FormattableAnyCalendarKind::Buddhist
+    );
+    assert_eq!(
+        FormattableAnyCalendarKind::from_preferences(locale!("en-TH").into()),
+        FormattableAnyCalendarKind::Buddhist
+    );
+    assert_eq!(
+        FormattableAnyCalendarKind::from_preferences(locale!("en-SA-u-ca-islamic").into()),
+        FormattableAnyCalendarKind::HijriUmmAlQura
+    );
```

Comment:
Nit: Add a test for `Hijri(None)` in a different region, and make it fall back to e.g. `HijriTabularTypeIIThursday`

Reply:
Ok, it's an easy thing to change now that the fallback algorithm is fixed